autoload/magit/utils.vim

```diff
@@ -20,12 +20,43 @@ function! magit#utils#is_submodule(dirname)
 	return ( index(s:submodule_list, a:dirname) != -1 )
 endfunction
 
-" s:magit_cd_cmd: plugin variable to choose lcd/cd command, 'lcd' if exists,
-" 'cd' otherwise
-let s:magit_cd_cmd = exists('*haslocaldir') && haslocaldir() ? 'lcd ' : 'cd '
-" magit#utils#lcd: helper function to lcd. use cd if lcd doesn't exists
-function! magit#utils#lcd(dir)
-	execute s:magit_cd_cmd . a:dir
+" magit#utils#chdir will change the directory respecting
+" local/tab-local/global directory settings.
+function! magit#utils#chdir(dir)
+  " This is a dirty hack to fix tcd breakages on neovim.
+  " Future work should be based on nvim API and has a `if has('nvim')` check.
+  if has('nvim')
+    let chdir = haslocaldir() ? 'lcd' : haslocaldir(-1, 0) ? 'tcd' : 'cd'
+  else
+    let chdir = exists('*haslocaldir') && haslocaldir() ? 'lcd' : 'cd'
+  endif
+  execute chdir . ' ' . a:dir
+endfunction
+
+" magit#utils#pushd_root goes to git root directory, saving current directory
+" prior to changing the cwd.
+function! magit#utils#pushd_root()
```

Comment:
`pushd` and and `popd` are not reentrant. Please add a counter in `pushd`, such as:
```
function pushd()
  b:pushd_cnt++
  if ( b:pushd_cnt <= 0 )
    raise error
  else if ( b:pushd_cnt == 1)
    b:_vimagit_cwd = cwd
    chdir root

function popd()
  b:pushd_cnt--
  if ( b:pushd_cnt < 0 )
    raise error
  else if ( b:pushd_cnt == 0 )
    chdir b:_vimagit_cwd
```